### PR TITLE
Add more type hints

### PR DESCRIPTION
**Pull request by dcermak**
Changes proposed in this pull request:
* Add type hints to more modules
* Utilize a `TypedDict` to better indicate required keys in `ContainerImageOCI.oci_config`

### CODE REVIEW

**Comment by dcermak** [comment]
The addition of `TypedDict` adds another requirement for Python <= 3.8: `typing_extensions`. If that is too much, then I can drop that addition as well.

**Comment by schaefi** [comment]
I like the changes :+1: , not sure about the py >= 3.8 (I think that's what you meant) requirement. For SLE development we need 3.6 and there is a fork of kiwi to maintain kiwi in SLE exactly for reasons like this. Of course it makes maintenance harder. 

I think @Conan-Kudo will object to the change because packaging for Fedora/CentOS/RHEL is based on the pypi upstream tarball and of the python doesn't match they would need to patch out things which is also hard to maintain.

I think we should discuss it either on matrix or in a call

**Comment by dcermak** [comment]
> not sure about the py >= 3.8 (I think that's what you meant) requirement

You don't need python 3.8 or later. Kiwi still works with 3.6+, but for 3.6 & 3.7 you need the typing_extensions python module as well to import `TypedDict`. I believe it is packaged everywhere, so I don't think it should be a blocker, but it would require updating the spec.

**Review by Conan-Kudo — CHANGES_REQUESTED** [review]

File: .virtualenv.requirements.txt [hunk 1]
```diff
@@ -23,3 +23,5 @@ xmltodict
 
 # json support that also knows about tuples
 simplejson
+
+typing_extensions; python_version < '3.8'
```
[the remark this answers is not in the document]
Comment (dcermak) [reply .virtualenv.requirements.txt:27]: Fixed & checked locally that it does the right thing^:tm:

**Review by Conan-Kudo — DISMISSED** [review]

File: .virtualenv.requirements.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (dcermak) [reply .virtualenv.requirements.txt:27]: I built the release tarball and pip installed it

**Review by Conan-Kudo — APPROVED** [review]

**Pull request merged**